docs/tutorial/frontend-tut.md

```diff
@@ -1,93 +1,158 @@
 # Building a Frontend for Calyx
 
-In this tutorial, we're going to build a compiler for a small language called MrXL.
+> This tutorial assumes that you have already worked through the [Calyx tutorial][calyx-tut]. You won't get very much out of it by itself!
 
-## MrXL Overview
-MrXL provides constructs to create arrays, and perform `map` and `reduce` operations on those arrays. Here's an example of a dot product implementation in MrXL:
+In the [Calyx tutorial][calyx-tut] you wrote Calyx by hand.
+That is (probably) a good way to build character, but it's no way to live.
+In practice, you want a *frontend* that *compiles* to Calyx.
+
+This allows you to:
+1. Support domain-specific commands that Calyx does not.
+2. Automatically generate some of the kludge that Calyx requires.
+
+In this tutorial, we're going to learn all about this by building a compiler for a toy language.
+Meet MrXL.
+
+# MrXL Overview
+
+MrXL lets you define arrays and registers and then perform `map` and `reduce` operations.
+
+## A tiny example
+
+Here's a MrXL program in all its glory:
 
 ```
-{{#include ../../frontends/mrxl/test/dot.mrxl}}
+{{#include ../../frontends/mrxl/test/sos.mrxl}}
 ```
 
-We define the interface of program by specifying `input` and `output` arrays.
-Input arrays have their values populated by an external harness while the output arrays must be computed using the program.
+The program is short enough for us to pick apart line by line:
```

Comment:
Perhaps it would be a good idea to state the maybe-obvious fact that it computes the sum or squares of an input array.